### PR TITLE
Add logger_snapshot_event_type flag for snapshot events

### PR DESCRIPTION
**Pull request by theopolis**
This implements #3701.

```
» cat ./build/example.conf 
{
  "options": {
    "logger_snapshot_event_type": "true"
  },
  "schedule": {
    "3groups": {
      "query": "select * from groups limit 3",
      "interval": 5,
      "snapshot": "true"
    }
  }
}
» ./build/linux/osquery/osqueryd --ephemeral --disable_database --logger_plugin=stdout --config_path ./build/example.conf
{"name":"3groups","hostIdentifier":"maverics","calendarTime":"Mon Oct 16 05:50:00 2017 UTC","unixTime":"1508133000","epoch":"0","counter":"0","columns":{"gid":"0","gid_signed":"0","groupname":"root"},"action":"snapshot"}
{"name":"3groups","hostIdentifier":"maverics","calendarTime":"Mon Oct 16 05:50:00 2017 UTC","unixTime":"1508133000","epoch":"0","counter":"0","columns":{"gid":"1","gid_signed":"1","groupname":"daemon"},"action":"snapshot"}
{"name":"3groups","hostIdentifier":"maverics","calendarTime":"Mon Oct 16 05:50:00 2017 UTC","unixTime":"1508133000","epoch":"0","counter":"0","columns":{"gid":"2","gid_signed":"2","groupname":"bin"},"action":"snapshot"}
severity=0 location=events.cpp:824 message=Event publisher not enabled: audit: Publisher disabled via configuration
severity=0 location=events.cpp:824 message=Event publisher not enabled: syslog: Publisher disabled via configuration
severity=0 location=scheduler.cpp:75 message=Executing scheduled query 3groups: select * from groups limit 3
{"name":"3groups","hostIdentifier":"maverics","calendarTime":"Mon Oct 16 05:50:05 2017 UTC","unixTime":"1508133005","epoch":"0","counter":"0","columns":{"gid":"0","gid_signed":"0","groupname":"root"},"action":"snapshot"}
{"name":"3groups","hostIdentifier":"maverics","calendarTime":"Mon Oct 16 05:50:05 2017 UTC","unixTime":"1508133005","epoch":"0","counter":"0","columns":{"gid":"1","gid_signed":"1","groupname":"daemon"},"action":"snapshot"}
{"name":"3groups","hostIdentifier":"maverics","calendarTime":"Mon Oct 16 05:50:05 2017 UTC","unixTime":"1508133005","epoch":"0","counter":"0","columns":{"gid":"2","gid_signed":"2","groupname":"bin"},"action":"snapshot"}
severity=0 location=scheduler.cpp:75 message=Executing scheduled query 3groups: select * from groups limit 3
```

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit 0418840e36ca23aaac398d44a11947bcbece64b1 (Job results: [939](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/939/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 0418840e36ca23aaac398d44a11947bcbece64b1 (Job results: [2127](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2127/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 0418840e36ca23aaac398d44a11947bcbece64b1 (Job results: [6022](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6022/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 0418840e36ca23aaac398d44a11947bcbece64b1 (Job results: [6023](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6023/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 0418840e36ca23aaac398d44a11947bcbece64b1 (Job results: [940](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/940/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 0418840e36ca23aaac398d44a11947bcbece64b1 (Job results: [2128](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2128/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit c80888b3f31a056aa12751c2011d15bb96674bb2 (Job results: [2900](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2900/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit c80888b3f31a056aa12751c2011d15bb96674bb2 (Job results: [2901](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2901/)) failed one or more tests (**Windows**).

File: osquery/core/query.cpp [hunk 1]
```diff
@@ -506,26 +506,71 @@ inline void getLegacyFieldsAndDecorations(const pt::ptree& tree,
...
   addLegacyFieldsAndDecorations(item, tree);
   return Status(0, "OK");
 }
 
+static inline Status serializeEvent(const QueryLogItem& item,
+                                    const pt::ptree& event,
+                                    pt::ptree& tree) {
+  addLegacyFieldsAndDecorations(item, tree);
+  pt::ptree columns;
+  for (auto& i : event) {
```
Comment (obelisk) [inline osquery/core/query.cpp]: `const`?

File: osquery/core/query.cpp [hunk 2]
```diff
@@ -506,26 +506,71 @@ inline void getLegacyFieldsAndDecorations(const pt::ptree& tree,
...
+
+Status serializeQueryLogItemAsEvents(const QueryLogItem& item,
+                                     pt::ptree& tree) {
+  pt::ptree results;
+  if (item.results.added.size() > 0 || item.results.removed.size() > 0) {
+    auto status = serializeDiffResults(item.results, results);
+    if (!status.ok()) {
+      return status;
+    }
+  } else if (item.snapshot_results.size() > 0) {
```
Comment (obelisk) [inline osquery/core/query.cpp]: Prefer `!item.snapshot_results.empty()`

File: osquery/core/query.cpp [hunk 3]
```diff
@@ -506,26 +506,71 @@ inline void getLegacyFieldsAndDecorations(const pt::ptree& tree,
...
+    if (!status.ok()) {
+      return status;
+    }
+    results.add_child("snapshot", snapshot_results);
+  } else {
+    // This error case may also be represented in serializeQueryLogItem.
+    return Status(1, "No diff results or snapshot results");
+  }
+
+  for (auto& action : results) {
```
Comment (obelisk) [inline osquery/core/query.cpp]: Can these be `const`s?

File: osquery/core/query.cpp [hunk 4]
```diff
@@ -506,26 +506,71 @@ inline void getLegacyFieldsAndDecorations(const pt::ptree& tree,
...
+  }
+
+  tree.add_child("columns", columns);
+  return Status(0, "OK");
+}
+
+Status serializeQueryLogItemAsEvents(const QueryLogItem& item,
+                                     pt::ptree& tree) {
+  pt::ptree results;
+  if (item.results.added.size() > 0 || item.results.removed.size() > 0) {
```
Comment (obelisk) [inline osquery/core/query.cpp]: If possible, prefer not empty

**Pull request merged**